sdk/src/react/ui/modals/SellModal/internal/context.ts

```diff
@@ -72,82 +73,96 @@ export function useSellModalContext() {
 	const [pendingFee] = useWaasFeeOptions();
 	const isSponsored = (pendingFee?.options?.length ?? -1) === 0;
 
-	const steps = [];
+	const steps: SellModalSteps = {} as SellModalSteps;
 
-	// Step 1: WaaS fee selection if needed, TODO: this need to be refactored to be headless and we need to take care of fees from both transactions
 	if (isWaaS) {
 		const feeSelected = isSponsored || !!waas.selectedFeeOption;
-		steps.push({
-			id: 'waasFee' satisfies SellStepId,
+
+		steps.fee = {
 			label: 'Select Fee',
-			status: feeSelected ? 'success' : 'idle',
-			isPending: false,
-			isSuccess: feeSelected,
-			isError: false, // TODO: Fee loading errors not accessible from useWaasFeeOptions
-			run: () => selectWaasFeeOptionsStore.send({ type: 'show' }),
-		});
+			status: feeSelected ? 'complete' : waas.isVisible ? 'selecting' : 'idle',
+			isSponsored,
+			isSelecting: waas.isVisible,
+			selectedOption: waas.selectedFeeOption,
+			show: () => selectWaasFeeOptionsStore.send({ type: 'show' }),
+			cancel: () => selectWaasFeeOptionsStore.send({ type: 'hide' }),
+		};
 	}
 
-	// Step 2: Approve (if needed)
-	if (sellSteps.data?.approveStep && !approve.isSuccess) {
-		steps.push({
-			id: 'approve' satisfies SellStepId,
+	if (sellSteps.data?.approveStep) {
+		const approvalGuard = createApprovalGuard({
+			isFormValid: true,
+			txReady: !!sellSteps.data?.approveStep,
+			walletConnected: !!address,
+		});
+		const guardResult = approvalGuard();
+
+		const approveTransactionHash =
+			approve.data &&
+			'type' in approve.data &&
+			approve.data.type === 'transaction'
+				? approve.data.hash
+				: undefined;
+
+		steps.approval = {
 			label: 'Approve Token',
-			status: approve.status,
+			status: approve.isSuccess
+				? 'complete'
+				: approve.isPending
+					? 'pending'
+					: approve.error
+						? 'error'
+						: 'idle',
 			isPending: approve.isPending,
-			isSuccess: approve.isSuccess,
-			isError: !!approve.error,
-			run: () => approve.mutate(),
-		});
+			isComplete: approve.isSuccess,
+			isDisabled: !guardResult.canProceed,
+			disabledReason: guardResult.reason || null,
+			error: approve.error,
+			canExecute: guardResult.canProceed,
+			result: approveTransactionHash
+				? { type: 'transaction', hash: approveTransactionHash }
+				: null,
+			execute: async () => approve.mutate(),
+			reset: () => {},
+		};
 	}
 
-	// Step 3: Sell
-	// TODO: sell step never completes here, it completes via the success callback, we need to change this
-	steps.push({
-		id: 'sell' satisfies SellStepId,
-		label: 'Accept Offer',
-		status: sell.status,
-		isPending: sell.isPending,
-		isSuccess: sell.isSuccess,
-		isError: !!sell.error,
-		run: () => sell.mutate(),
+	const sellGuard = createFinalTransactionGuard({
```

Comment:
FinalTransaction sounds strange to me, does this makes sense? 👇🏼

```suggestion
	const sellGuard = createSaleTransactionGuard({
```